module/zfs/metaslab.c

```diff
@@ -2598,8 +2881,145 @@ metaslab_group_alloc_verify(spa_t *spa, const blkptr_t *bp, void *tag)
 #endif
 }
 
+/*
+ * block_alloc_stats() is common to metaslab layer and zdb (for auditing).
+ */
+static void
+block_alloc_stats(ms_alloc_stats_t *mas, int64_t asize, block_context_t *blkctx)
+{
+	int64_t psize;
+	int h, count;
+
+	if (asize < 0) {
+		count = -1;
+		psize = -((int64_t)blkctx->bc_psize);
+	} else {
+		count = 1;
+		psize = blkctx->bc_psize;
+	}
+
+	/*
+	 * Gang child members are not tracked
+	 */
+	if (blkctx->bc_child_gang)
+		return;
+
+	/*
+	 * transient zil blocks are never tracked
+	 */
+	if (DMU_OT_IS_ZIL(blkctx->bc_type) || blkctx->bc_type == DMU_OT_NONE)
+		return;
+
+	if (DMU_OT_IS_DDT(blkctx->bc_type)) {
+		mas->mas_dedup_bytes += psize;
+		mas->mas_dedup_blocks += count;
+		return;
+	}
+
+	/* File blocks */
+	if (DMU_OT_IS_FILE(blkctx->bc_type)) {
+		if (blkctx->bc_level == 0) {
+			mas->mas_file_bytes += psize;
+			mas->mas_file_blocks += count;
+
+			/* File small blocks (up to 128K) */
+			if (blkctx->bc_psize <= MS_SPECIAL_SMALLBLK_BOUNDS) {
+				h = highbit64(blkctx->bc_psize) -
+				    SPA_MINBLOCKSHIFT - 1;
+				if (h < CAS_HISTOGRAM_SIZE)
+					mas->mas_smallblk_histogram[h] += count;
+			}
+		} else {
+			mas->mas_file_indirect_bytes += psize;
+			mas->mas_file_indirect_blocks += count;
+		}
+		return;
+	}
+
+	/* zvol blocks */
+	if (blkctx->bc_type == DMU_OT_ZVOL) {
+		if (blkctx->bc_level == 0) {
+			mas->mas_zvol_bytes += psize;
+			mas->mas_zvol_blocks += count;
+		} else {
+			mas->mas_zvol_indirect_bytes += psize;
+			mas->mas_zvol_indirect_blocks += count;
+		}
+		return;
+	}
+
+	if (blkctx->bc_type == DMU_OT_DNODE) {
+		mas->mas_dnode_bytes += psize;
+		mas->mas_dnode_blocks += count;
+		return;
+	}
+
+	/* remaining metadata blocks */
+	if (DMU_OT_IS_METADATA(blkctx->bc_type) || blkctx->bc_level > 0) {
+		h = highbit64(blkctx->bc_psize) - SPA_MINBLOCKSHIFT - 1;
+		if (h < CAS_HISTOGRAM_SIZE)
+			mas->mas_metadata_histogram[h] += count;
+
+		return;
+	}
+}
+
+#ifndef _KERNEL
+/*
+ * A zdb(1M) block audit uses this to validate the on-disk allocation stats
+ */
```

Comment:
Could we instead hack up zdb to call metaslab_block_track(), and get the stats from ms_alloc_stats[]?  That would be more similar to what we do for zdb's other space accounting, I think.  If not, then let's rewrite metaslab_block_track() to call metaslab_class_block_track(), to avoid the code duplication.